### PR TITLE
Optimize bit operation in vm.c

### PR DESCRIPTION
**Pull request by lygstate**
JerryScript-DCO-1.0-Signed-off-by: Yonggang Luo luoyonggang@gmail.com

### CODE REVIEW

**Comment by rerobika** [comment]
Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 10992 -> 10992 : 0.000% | 13800 -> 13800 : 0.000% | 1.144 -> 1.146 : -0.231% | 
3d-morph.js | 289232 -> 289232 : 0.000% | 1576 -> 1576 : 0.000% | 1.337 -> 1.328 : +0.690% | 
3d-raytrace.js | 117656 -> 117656 : 0.000% | 2540 -> 2540 : 0.000% | 1.336 -> 1.336 : -0.008% | 
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3432 -> 3432 : 0.000% | 0.777 -> 0.778 : -0.116% | 
access-fannkuch.js | 1792 -> 1792 : 0.000% | 1648 -> 1648 : 0.000% | 2.564 -> 2.565 : -0.032% | 
access-nbody.js | 4600 -> 4600 : 0.000% | 1736 -> 1736 : 0.000% | 1.533 -> 1.532 : +0.013% | 
bitops-3bit-bits-in-byte.js | 1024 -> 1024 : 0.000% | 1592 -> 1592 : 0.000% | 0.716 -> 0.714 : +0.338% | 
bitops-bits-in-byte.js | 984 -> 984 : 0.000% | 1592 -> 1592 : 0.000% | 1.008 -> 1.012 : -0.462% | 
bitops-bitwise-and.js | 752 -> 752 : 0.000% | 1344 -> 1344 : 0.000% | 1.225 -> 1.225 : +0.055% | 
bitops-nsieve-bits.js | 61184 -> 61184 : 0.000% | 1592 -> 1592 : 0.000% | 1.498 -> 1.499 : -0.101% | 
controlflow-recursive.js | 1288 -> 1288 : 0.000% | 61584 -> 61584 : 0.000% | 0.528 -> 0.530 : -0.356% | 
crypto-aes.js | 28040 -> 28040 : 0.000% | 2288 -> 2288 : 0.000% | 0.884 -> 0.882 : +0.178% | 
crypto-md5.js | 70968 -> 70968 : 0.000% | 1704 -> 1704 : 0.000% | 0.747 -> 0.744 : +0.294% | 
crypto-sha1.js | 46328 -> 46328 : 0.000% | 1720 -> 1720 : 0.000% | 0.761 -> 0.761 : +0.031% | 
crypto.js | 51704 -> 51704 : 0.000% | 3156 -> 3156 : 0.000% | 8.707 -> 8.737 : -0.342% | 
date-format-tofte.js | 8952 -> 8952 : 0.000% | 2008 -> 2008 : 0.000% | 1.214 -> 1.224 : -0.761% | 
date-format-xparb.js | 15144 -> 15144 : 0.000% | 2552 -> 2552 : 0.000% | 1.632 -> 1.637 : -0.338% | 
deltablue.js | 455544 -> 455544 : 0.000% | 3540 -> 3540 : 0.000% | 5.420 -> 5.426 : -0.125% | 
math-cordic.js | 2048 -> 2048 : 0.000% | 1592 -> 1592 : 0.000% | 1.716 -> 1.710 : +0.318% | 
math-partial-sums.js | 1512 -> 1512 : 0.000% | 1568 -> 1568 : 0.000% | 1.151 -> 1.149 : +0.162% | 
math-spectral-norm.js | 3264 -> 3264 : 0.000% | 1592 -> 1592 : 0.000% | 0.740 -> 0.738 : +0.323% | 
raytrace.js | 22424 -> 22424 : 0.000% | 5000 -> 5000 : 0.000% | 2.640 -> 2.635 : +0.211% | 
richards.js | 8568 -> 8568 : 0.000% | 2140 -> 2140 : 0.000% | 0.310 -> 0.310 : +0.010% | 
string-base64.js | 90104 -> 90104 : 0.000% | 1640 -> 1640 : 0.000% | 1.421 -> 1.418 : +0.228% | 
string-fasta.js | 4400 -> 4400 : 0.000% | 1576 -> 1576 : 0.000% | 3.452 -> 3.451 : +0.026% | 
Geometric mean: | 0.000% | 0.000% | +0.001% | 

Binary (bytes) | master(3df7998084) | patch(98087ab736) | Diff |
----: | ----: | ----: | ----: |
size | 226976 | 226976 | 0 bytes |
.rodata | 20095 | 20095 | 0 bytes |
.dynstr | 446 | 446 | 0 bytes |
.rel.plt | 456 | 456 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 1008 | 1008 | 0 bytes |
.gnu.hash | 464 | 464 | 0 bytes |
.text | 200040 | 200040 | 0 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 244 | 244 | 0 bytes |
.plt | 716 | 716 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 64 | 64 | 0 bytes |
.bss | 1575096 | 1575096 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 408 | 408 | 0 bytes |
.gnu.version | 126 | 126 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 248 | 248 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |

I see no significant improvement, mostly noise.

File: jerry-core/vm/vm.c [hunk 1]
```diff
@@ -3312,7 +3312,7 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
 
           if (ecma_is_value_integer_number (left_value))
           {
-            *stack_top_p++ = (~left_value) & (ecma_value_t) (~ECMA_DIRECT_TYPE_MASK);
+            *stack_top_p++ = 0xFFFFFFF0 ^ left_value;
```
Comment (zherczeg) [inline jerry-core/vm/vm.c:3315]: Isn't this is 0xFFFFFFF0 == (~ECMA_DIRECT_TYPE_MASK)? I would prefer some symbols here.

**Review by zherczeg — CHANGES_REQUESTED** [review]
These are good changes, regardless the improvement.

**Review by rerobika — APPROVED** [review]

**Review by zherczeg — APPROVED** [review]
LGTM

**Pull request merged**